src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp

```diff
@@ -202,6 +202,141 @@ void runtime::gpu::CudaKernelBuilder::get_ew_collective_op(
+    writer.block_begin();
+    {
+        writer << "Entry t = a;\n";
+        writer << "a = b;\n";
+        writer << "b = t;\n";
+    }
+    writer.block_end();
+    writer << "__device__ void heapify(Entry *heap, size_t heap_size, size_t idx)\n";
+    writer.block_begin();
+    {
+        writer << "size_t largest = idx;\n";
+        writer << "size_t left = (idx << 1) + 1;\n";
+        writer << "size_t right = (idx + 1) << 1;\n";
+        std::string g_op = ((compute_max) ? ">" : "<");
+        writer << "if (left < heap_size && heap[left].get_value() " << g_op
+               << " heap[largest].get_value())\n";
+        writer.block_begin();
+        {
+            writer << "largest = left;\n";
+        }
+        writer.block_end();
+        writer << "if (right < heap_size && heap[right].get_value() " << g_op
+               << " heap[largest].get_value())\n";
+        writer.block_begin();
+        {
+            writer << "largest = right;\n";
+        }
+        writer.block_end();
+        writer << "if (largest != idx)\n";
+        writer.block_begin();
+        {
+            writer << "swap(heap[largest], heap[idx]);\n";
+            writer << "heapify(heap, heap_size, largest);\n";
+        }
+        writer.block_end();
+    }
+    writer.block_end();
+    writer << "__device__ void create_and_build(Entry *entry, size_t size)\n";
+    writer.block_begin();
+    {
+        writer << "for (int i = (size-2) / 2; i >= 0; --i)\n";
+        writer.block_begin();
+        {
+            writer << "heapify(entry, size, i);\n";
+        }
+        writer.block_end();
+    }
+    writer.block_end();
+
+    if (use_malloc)
+    {
+        writer << "extern \"C\" __global__ void cuda_" << name << "(";
+        writer << dtypes[0] << "* in, " << dtypes[1] << "* out_id, " << dtypes[2] << "* out_val, "
+               << " Entry* entry,";
+        writer << "size_t num_cols, size_t topk_k)\n";
+    }
+    else
+    {
+        writer << "extern \"C\" __global__ void cuda_" << name << "(";
+        writer << dtypes[0] << "* in, " << dtypes[1] << "* out_id, " << dtypes[2] << "* out_val, ";
+        writer << " size_t num_cols, size_t topk_k)\n";
+    }
+    writer.block_begin();
+    {
+        writer << "in = in + blockIdx.x * num_cols;\n";
+        if (use_malloc)
+        {
+            writer << "entry = entry + blockIdx.x * num_cols;\n";
+        }
+        writer << "out_id = out_id + blockIdx.x * topk_k;\n";
+        writer << "out_val = out_val + blockIdx.x * topk_k;\n";
+        if (!use_malloc)
+        {
+            writer << "extern __shared__ Entry entry[];\n";
+        }
+
+        writer << "for (size_t i = threadIdx.x; i < num_cols; i += blockDim.x)\n";
+        writer.block_begin();
+        {
+            writer << "entry[i].set_value(in[i]);\n";
+            writer << "entry[i].set_index(i);\n";
+        }
+        writer.block_end();
+
+        writer << "__syncthreads();\n";
+
+        writer << "if (threadIdx.x == 0)\n";
```

Comment:
Since this is roughly O(nlogn), the transpose (O(n^2)) will dominate the kernel and so doing a strided load and write will likely be better. tensorflow is ok with this because they rely mainly on a radix sort from cub (for n <=1000, when k=n or when k > 100). It would be good to get a serialized graph with topk in it to compare. but this is lower priority right now.

Reply:
Agreed!